supermarq-benchmarks/supermarq/qcvv/base_experiment.py

```diff
@@ -0,0 +1,589 @@
+        """
+        if self._results is None:
+            raise RuntimeError("No results to retrieve. The experiment has not been run.")
+
+        return self._results
+
+    @property
+    def samples(self) -> Sequence[Sample]:
+        """The samples generated during the experiment.
+
+        Raises:
+            RuntimeError: If no samples are available.
+        """
+        if self._samples is None:
+            raise RuntimeError("No samples to retrieve. The experiment has not been run.")
+
+        return self._samples
+
+    @property
+    def targets(self) -> frozenset[str]:
+        """Returns:
+        A set of the unique target that each sample was submitted to
+        """
+        return frozenset(sample.target for sample in self.samples)
+
+    ###################
+    # Private Methods #
+    ###################
+    @abstractmethod
+    def _build_circuits(
+        self,
+        num_circuits: int,
+        cycle_depths: Iterable[int],
+    ) -> Sequence[Sample]:
+        """Build a list of circuits required for the experiment. These circuits are stored in
+        :class:`Sample` objects along with any additional data that is needed during the analysis.
+
+        Args:
+            num_circuits: Number of circuits to generate.
+            cycle_depths: An iterable of the different cycle depths to use during the experiment.
+
+        Returns:
+           The list of experiment samples.
+        """
+
+    @staticmethod
+    def _interleave_op(
+        circuit: cirq.Circuit, operation: cirq.Operation, include_final: bool = False
+    ) -> cirq.Circuit:
+        """Interleave a given operation into a circuit.
+
+        Args:
+            circuit: The original circuit.
+            operation: The operation to interleave.
+            include_final: If True then the interleaving gate is also appended to
+                the end of the circuit.
+
+        Returns:
+            A copy of the original circuit with the provided gate interleaved.
+        """
+        operation = operation.with_tags("no_compile")
+        interleaved_circuit = circuit.copy()
+        interleaved_circuit.batch_insert(
+            [(k, operation) for k in range(len(circuit) - int(not include_final), 0, -1)]
+        )
+        return interleaved_circuit
+
+    def _process_device_counts(self, counts: dict[str, int]) -> dict[str, float]:
+        """Process the counts returned by the server into a dictionary of probabilities.
+
+        Args:
+            counts: A dictionary of the observed counts for each state in the computational basis.
+
+        Returns:
+            A dictionary of the probability of each state in the computational basis.
+        """
+        total = sum(counts.values())
+
+        probabilities = {
+            format(idx, f"0{self.num_qubits}b"): 0.0 for idx in range(2**self.num_qubits)
+        }
+
+        for key, count in counts.items():
+            probabilities[key] = count / total
```

Comment:
Because of they way these probability dictionaries are treated as records in a dataframe, to avoid NaNs I prefer enforcing that all the bitstrings are present in the dictionary.

Reply:
ok sounds good. in that case i think we also need to fill in the missing values in the simulator output?